Add discussion that concurrent access to the environment is unsafe

tbu- opened the pull request:
The bug report #27970 has existed for 8 years, the actual bug dates back to Rust pre-1.0. I documented it since it's in the interest of the user to be aware of it. The note can be removed once #27970 is fixed.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6561026380/job/17819866354) [(plain)](https://github.com/rust-lang/rust/commit/8436e8f35e6c48523f31a1f81658f5c5a26646fd/checks/17819866354/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_5a971fc6-bbde-45b6-a010-1526d2118d8e
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=pr_unsafe_env
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_5a971fc6-bbde-45b6-a010-1526d2118d8e
GITHUB_REF=refs/pull/116888/merge
GITHUB_REF_NAME=116888/merge
GITHUB_REF_PROTECTED=false
---
 Documenting std v0.0.0 (/checkout/library/std)
error: unresolved link to `std::net::ToSocketAddrs`
   --> library/std/src/env.rs:321:7
    |
321 | /// [`std::net::ToSocketAddrs`]. This is a bug
    |
    |
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
error: unresolved link to `std::net::ToSocketAddrs`
   --> library/std/src/env.rs:366:7
    |
    |
366 | /// [`std::net::ToSocketAddrs`]. This is a bug

error: could not document `std`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type dylib --crate-type rlib --crate-name std library/std/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/doc --cfg 'feature="addr2line"' --cfg 'feature="backtrace"' --cfg 'feature="compiler-builtins-c"' --cfg 'feature="gimli-symbolize"' --cfg 'feature="miniz_oxide"' --cfg 'feature="object"' --cfg 'feature="panic_unwind"' --cfg 'feature="std_detect_dlsym_getauxval"' --cfg 'feature="std_detect_file_io"' -Zunstable-options --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=01dc999398991194 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/release/deps --extern 'priv:addr2line=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libaddr2line-ac8dd5a11a6031bd.rmeta' --extern alloc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/liballoc-aabf62574aec6512.rmeta --extern 'priv:cfg_if=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libcfg_if-9c565ddf5a0048e0.rmeta' --extern 'priv:compiler_builtins=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libcompiler_builtins-192937f85280d0b5.rmeta' --extern core=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libcore-6f6461ce7e567ed5.rmeta --extern 'priv:hashbrown=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libhashbrown-ce7b17b3f1ccaeba.rmeta' --extern libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/liblibc-8937af5b522b0640.rmeta --extern 'priv:miniz_oxide=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libminiz_oxide-19f2a730eebfabbe.rmeta' --extern 'priv:object=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libobject-c7a9f31fc4eed475.rmeta' --extern 'priv:panic_abort=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libpanic_abort-9ce0bcf76e3e1af0.rmeta' --extern 'priv:panic_unwind=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libpanic_unwind-80572631ee3b437b.rmeta' --extern 'priv:rustc_demangle=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/librustc_demangle-ab192e1c5112009e.rmeta' --extern 'priv:std_detect=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libstd_detect-c66ae6db34967bbe.rmeta' --extern 'priv:unwind=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libunwind-2eb50eb34932a8ab.rmeta' -Z unstable-options --cfg=bootstrap -Csymbol-mangling-version=legacy -Zunstable-options -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(stdarch_intel_sde)' '--check-cfg=values(no_fp_fmt_parse)' '--check-cfg=values(no_global_oom_handling)' '--check-cfg=values(no_rc)' '--check-cfg=values(no_sync)' '--check-cfg=values(freebsd12)' '--check-cfg=values(freebsd13)' '--check-cfg=values(backtrace_in_libstd)' '--check-cfg=values(target_env,"libnx")' '--check-cfg=values(target_arch,"asmjs","spirv","nvptx","xtensa")' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.75.0-nightly (a02048ac2 2023-10-18)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' '-Zcrate-attr=warn(rust_2018_idioms)' -Z unstable-options --resource-suffix 1.75.0 --markdown-css rust.css --markdown-no-toc --index-page /checkout/src/doc/index.md --cfg backtrace_in_libstd` (exit status: 1)
Build completed unsuccessfully in 0:00:15
  local time: Wed Oct 18 12:56:32 UTC 2023
  network time: Wed, 18 Oct 2023 12:56:32 GMT
##[error]Process completed with exit code 1.
```

</details>

tbu- committed:
Add discussion that concurrent access to the environment is unsafe

The bug report #27970 has existed for 8 years, the actual bug dates back
to Rust pre-1.0. I documented it since it's in the interest of the user
to be aware of it. The note can be removed once #27970 is fixed.

Mark-Simulacrum commented on library/std/src/env.rs:
```diff
@@ -313,17 +313,26 @@ impl Error for VarError {
+/// behavior, for example in combination with DNS lookups from
+/// [`std::net::ToSocketAddrs`]. This is a bug
+/// ([rust#27970](https://github.com/rust-lang/rust/issues/27970)) and will be
+/// fixed in a future version of Rust. Additionally, extra care needs to be
```
It also seems to me that get_var has the same problems, right? It's reading data that might be changed by concurrent C code.

tbu- replied:
Since `getenv` is usually considered safe in a multithreaded C environment, but `setenv`/`putenv` is not, I do not think that it requires an extra note.

thomcc replied:
Yeah, C code designed for multithreaded use will almost never write to the env.

thomcc commented on library/std/src/env.rs:
```diff
@@ -313,17 +313,26 @@ impl Error for VarError {
+/// taken when auditing calls to unsafe external FFI functions to ensure that
+/// any external environment accesses are properly synchronized with accesses
+/// in Rust. Since Rust does not expose its environment lock directly, this
+/// means that all accesses to the environment must go through Rust's [`var`].
```
This is requesting something impossible in most cases.

tbu- replied:
Do you have a suggestion what to write there?

Yes, it's unreasonable, but that's the current state of the function AFAICT.

tbu- commented:
@rustbot ready

tbu- commented:
@Mark-Simulacrum I don't see any actionable feedback, why is it "waiting-on-author"?

Mark-Simulacrum commented:
> fixed in a future version of Rust

I think at minimum we need to remove this implication, this isn't directly fixable in a future version of Rust.

I'm going to also nominate for T-libs-api -- it's not clear to me if the framing here meets our current thinking. In particular,

> Note that while concurrent access to environment variables ought to be safe in Rust

My suspicion is that in hindsight these functions should have either not existed or not been safe. i.e., "ought to be safe" isn't accurate here -- usually we do not try to paper over platform unsoundness like this, especially if it's impossible (as is the case here).

Mark-Simulacrum commented:
(It also seems like this is just extending existing documentation -- i.e., the documentation already calls out this is an underlying unsafe API)

Amanieu commented:
We discussed this in the libs-api meeting. What we would like to see happen is that this function gets turned into an unsafe function for the 2024 edition, but on older editions only emits a warning if not inside an `unsafe` block (possibly using `deprecated_safe` #94978).

As such the documentation should be written as if this was an unsafe function, clearly explaining the pre-conditions that must be upheld:
- This function is safe to call in single-threaded programs.
- In multi-threaded programs, you must ensure that there are no other threads concurrently reading or writing the environment variables while calling this function through APIs other than the standard library (e.g. C code). You are responsible for figuring out how to achieve this, but we strongly suggest not calling `set_var` in multi-threaded programs at all.

ChrisDenton commented:
> In multi-threaded programs, you must ensure that there are no other threads concurrently reading or writing the environment variables while **calling this function through APIs other than the standard library** (e.g. C code).

It was also brought up that this isn't quite true currently as the rust standard library calls libc functions that may themselves read from the environment. So either those std functions need to be changed to take the environment lock or else we also need to warn against using std functions that may call out to libc.

tbu- committed:
Reformulate `std::env::{set,remove}_env` as safety note

tbu- commented:
> As such the documentation should be written as if this was an unsafe function, clearly explaining the pre-conditions that must be upheld: […]

Done.

Amanieu commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 2093d0c58e4508012439a55d6c5a52929d16748f has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 2093d0c58e4508012439a55d6c5a52929d16748f -->
<!-- homu: {"type":"Approved","sha":"2093d0c58e4508012439a55d6c5a52929d16748f","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

hkBst commented:
The suggested wording seems to imply that it is fine to call the functions in this module in parallel if you don't use any other ways to access the environment, presumably because these functions use some way to synchronize their accesses. It would be great if the text was more explicit about whether this is the case or not.

tbu- commented:
> The suggested wording seems to imply that it is fine to call the functions in this module in parallel if you don't use any other ways to access the environment, presumably because these functions use some way to synchronize their accesses.

You must also ensure that you don't call any Rust standard library function that reads from the environment (e.g. `ToSocketAddrs` does). It is not documented which Rust standard library function reads from the environment and can change from version to version.

You must also not call into any C libraries that might read from the environment. Most C libraries do not document whether or in which functions they access the environment, and it can usually also change from version to version.

In that way, you could say that your statement is technically correct. Practically speaking, since you cannot even guarantee what the Rust standard library does, you must only modify the environment variables when you're single-threaded.

> It would be great if the text was more explicit about whether this is the case or not.

I agree, and this is why I suggested removing the lock guarantees from the documentation: #125937. IMO unfortunately, this seems to be rejected. You could add a comment there if you think it's an improvement.
